core/payment/src/processor.rs

```diff
@@ -1,203 +1,288 @@
+                (Some(activity_id), None) => activity_payments.push(ActivityPayment {
+                    activity_id,
+                    amount,
+                    allocation_id: Some(order.allocation_id.clone()),
+                }),
+                (None, Some(agreement_id)) => agreement_payments.push(AgreementPayment {
+                    agreement_id,
+                    amount,
+                    allocation_id: Some(order.allocation_id.clone()),
+                }),
+                _ => return NotifyPaymentError::invalid_order(&order),
+            }
+        }
+
+        // FIXME: This is a hack. Payment orders realized by a single transaction are not guaranteed
+        //        to have the same payer and payee IDs. Fixing this requires a major redesign of the
+        //        data model. Payments can no longer by assigned to a single payer and payee.
+        let payer_id = orders.get(0).unwrap().payer_id;
+        let payee_id = orders.get(0).unwrap().payee_id;
+
+        let payment_dao: PaymentDao = self.db_executor.as_dao();
+        let payment_id = payment_dao
+            .create_new(
+                payer_id,
+                payee_id,
+                payer_addr,
+                payee_addr,
+                payment_platform,
+                msg.amount,
+                msg.confirmation.confirmation,
+                activity_payments,
+                agreement_payments,
+            )
+            .await?;
+
+        let payment = payment_dao.get(payment_id, payer_id).await?.unwrap();
+        let msg = SendPayment(payment);
+        ya_net::from(payer_id)
+            .to(payee_id)
+            .service(BUS_ID)
+            .call(msg)
+            .await??;
+
+        // TODO: Implement re-sending mechanism in case SendPayment fails
+
+        Ok(())
+    }
+
+    pub async fn schedule_payment(&self, msg: SchedulePayment) -> Result<(), SchedulePaymentError> {
+        let amount = msg.amount.clone();
+        let driver = self
+            .driver(&msg.payment_platform, &msg.payer_addr, AccountMode::SEND)
+            .await?;
+        let order_id = driver_endpoint(&driver)
             .send(driver::SchedulePayment::new(
-                document_id,
                 amount,
                 msg.payer_addr.clone(),
                 msg.payee_addr.clone(),
                 msg.due_date.clone(),
             ))
-            .await
-            .unwrap()
-            .unwrap();
+            .await??;
 
-        let processor = self.clone();
-        tokio::task::spawn_local(async move {
-            processor.process_payment(msg).await;
-        });
+        self.db_executor
+            .as_dao::<OrderDao>()
+            .create(msg, order_id, driver)
+            .await?;
 
         Ok(())
     }
 
-    pub async fn verify_payment(&self, payment: Payment) -> Result<(), Error> {
+    pub async fn verify_payment(&self, payment: Payment) -> Result<(), VerifyPaymentError> {
```

Comment:
Such a long function... This is a reason why I like to split multiple checks into single functions - it is easy to test them ;)

Reply:
That's a valid point, but this function wasn't introduced in this PR so I don't think it should be fixed in it. I will add a TODO comment and can open a separate PR later.

Reply:
Ok